staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -220,18 +238,59 @@ func (s *Controller) Run(stopCh <-chan struct{}, workers int) {
 		go wait.Until(s.worker, time.Second, stopCh)
 	}
 
-	go wait.Until(s.nodeSyncLoop, nodeSyncPeriod, stopCh)
+	go s.nodeSyncLoop(workers)
+	go wait.Until(s.triggerNodeSync, nodeSyncPeriod, stopCh)
 
 	<-stopCh
 }
 
+// triggerNodeSync triggers a nodeSync asynchronously
+func (s *Controller) triggerNodeSync() {
+	s.nodeSyncLock.Lock()
+	defer s.nodeSyncLock.Unlock()
+	newHosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to retrieve current set of nodes from node lister: %v", err))
+		// if node list cannot be retrieve, trigger full node sync to be safe.
+		s.needFullSync = true
```

Comment:
This should only happen on corrupted data in the cache. So yeah - we can do whatever here.